gk/main.c

```diff
@@ -16,22 +16,404 @@
  */
 
+#include <stdbool.h>
+
+#include <rte_ip.h>
+#include <rte_log.h>
+#include <rte_hash.h>
+#include <rte_lcore.h>
 #include <rte_ethdev.h>
+#include <rte_memcpy.h>
+#include <rte_cycles.h>
 #include <rte_malloc.h>
 
 #include "gatekeeper_gk.h"
 #include "gatekeeper_main.h"
 #include "gatekeeper_net.h"
 
+#ifdef RTE_MACHINE_CPUFLAG_SSE4_2
+#include <rte_hash_crc.h>
+#define DEFAULT_HASH_FUNC       rte_hash_crc
+#else
+#include <rte_jhash.h>
+#define DEFAULT_HASH_FUNC       rte_jhash
+#endif
+
+#define	START_PRIORITY		 (38)
+/* Set @START_ALLOWANCE as the double size of a large DNS reply. */
+#define	START_ALLOWANCE		 (8)
+
 #define GATEKEEPER_MAX_PKT_BURST (32)
 
+/*
+ * A flow entry can be in one of three states:
+ * request, granted, or declined.
+ */
+enum gk_flow_state { GK_REQUEST, GK_GRANTED, GK_DECLINED };
+
+struct ip_flow {
+	/* IPv4 or IPv6. */
+	uint16_t proto;
+
+	union {
+		struct {
+			uint32_t src;
+			uint32_t dst;
+		} v4;
+
+		struct ipv6_flow {
+			uint8_t src[16];
+			uint8_t dst[16];
+		} v6;
+	} f;
+};
+
+/*
+ * XXX Stores information about the packet, add new fields if necessary.
+ */
+struct ipacket {
+	struct ip_flow  flow;
+	struct rte_mbuf *pkt;
+};
+
+struct flow_entry {
+	/* IP flow information. */
+	struct ip_flow flow;
+
+	/* The state of the entry. */
+	unsigned state : 2;
+
+	union {
+		struct {
+			/* The time the last packet of the entry was seen. */
+			uint64_t last_packet_seen_at;
+			/* 
+			 * The priority associated to
+			 * the last packet of the entry.
+			 */
+			uint8_t last_priority;
+			/* 
+			 * The number of packets that the entry is allowed
+			 * to send with @last_priority without waiting
+			 * the amount of time necessary to be granted
+			 * @last_priority.
+			 */
+			uint8_t allowance;
+			/* 
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+		} request;
+
+		struct {
+			/* When the granted capability expires. */
+			uint64_t cap_expire_at;
+			/* When @budget_byte is reset. */
+			uint64_t budget_renew_at;
+			/* 
+			 * When @budget_byte is reset, reset it to
+			 * @tx_rate_kb_cycle * 1024 bytes.
+			 */
+			int tx_rate_kb_cycle;
+			/* How many bytes @src can still send in current cycle. */
+			int budget_byte;
+			/*
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+			/* When GK should send the next renewal to @grantor_id. */
+			uint64_t send_next_renewal_at;
+			/*
+			 * How many cycles (unit) GK must wait before
+			 * sending the next capability renewal request.
+			 */
+			uint64_t renewal_step_cycle;
+		} granted;
+
+		struct {
+			/*
+			 * When the punishment (i.e. the declined capability)
+			 * expires.
+			 */
+			uint64_t expire_at;
+		} declined;
+	} u;
+};
+
+static uint8_t
+integer_log_base_2(uint64_t delta_time)
+{
+	return (8 * sizeof(uint64_t) - __builtin_clzll((delta_time)) - 1);
```

Comment:
Thanks Michel for the detailed explanation. I have checked ``stdint.h``, and this works well.